[blackjack - Step2] Submit SebasNadu's mission

SebasNadu opened the pull request:
Hi Wordbe! 👋

This PR adds betting mechanics and improves class design in the Blackjack game. I also made several refactorings to improve readability, encapsulation, and separation of concerns.

I didn’t go forward with the state pattern this time, I’ve parked it on my TODO list so I can continue with the course flow. That said, I’d love to hear your thoughts and learn more from your perspective as always!

### Betting Mechanics:
* Introduced a new `Chips` value class to manage betting amounts, ensuring immutability and providing operations like addition, subtraction, and multiplication for bet calculations (`src/main/kotlin/blackjack/model/Chips.kt`).
* Added a `placeBets` step in the game flow to allow players to place bets before the game starts.
* Updated the `Result` enum to include a `BLACKJACK` result for handling blackjack-specific outcomes.

### Class Design Improvements:
* Updated the `Card` class to implement `Comparable`, enabling sorting of cards by rank, after the first two cards.
* Added a `sortCards` method in the `Hand` class and a corresponding `sortHand` method in the `Participant` class to sort cards for better readability (`src/main/kotlin/blackjack/model/card/Hand.kt`.

### Codebase Simplification:
* Removed redundant `toString()` implementations in several classes, delegating responsibility for string representation to the view layer.
* Changed internal methods in `BlackJackGame` to private to encapsulate functionality and improve code safety, when they are not using for testing reasons.

### View and Input Enhancements:
* Updated `InputView` to include a method for retrieving betting amounts and validating that the input is a positive number.
* Adjusted `OutputView` to display final earnings instead of results, showing profits for both the dealer and players.

### TODO
* Implement the state pattern in Participant and centralize the result and profit logic there.

### Example

<img width="491" height="765" alt="Screenshot 2025-07-14 at 15 20 01" src="https://github.com/user-attachments/assets/b15261f3-2a09-4016-b440-0ee1da34cb0d" />

SebasNadu committed:
refactor: update internal methods to private for encapsulation

SebasNadu committed:
refactor: remove redundant `toString` methods from models and update references

SebasNadu committed:
feat: add hand sorting functionality for participants and cards

SebasNadu committed:
docs: update TODO list with new features and modifications

SebasNadu committed:
feat: add `Money` value class, betting functionality for players and profit: Money property for Participant

SebasNadu committed:
feat: implement betting flow for players in BlackJack game

SebasNadu committed:
refactor: replace `Money` with `Chips` to improve naming consistency

SebasNadu committed:
feat: add profit calculation and BLACKJACK result handling for participants

SebasNadu committed:
refactor: update result display to show profits instead of results

SebasNadu committed:
docs: update README to reflect `Chips` value class implementation and completion

SebasNadu committed:
refactor: improve code readability and update `Card` to data class

- Reformatted code in `Player` for better readability.
- Made `Card` a data class to facility the tests with Card.
- Added missing newline and organized spacing in `Chips` for clarity.

SebasNadu committed:
refactor: rename variables and simplify `Deck` initialization for clarity

- Changed `_hand` to `participantHand` for better naming consistency.
- Renamed `lastResult` to `result` in `PlayerResultTracker` to align with usage.
- Simplified `Deck` initialization by introducing `initialCards` parameter and updating `generateADeck` method.
- Replaced redundant code in hand-related methods for participants.

SebasNadu committed:
test: add unit tests for `GameJudge` to validate evaluation and profit logic

- Added `GameJudgeTest` to test result evaluation logic for dealer and players.
- Verified correct application of profit, loss, and result updates in different scenarios.

SebasNadu committed:
refactor: make `result` property in `ResultTracker` an interface member

- Moved `result` property to `ResultTracker` interface to standardize implementation across subclasses.
- Updated `PlayerResultTracker` and `DealerResultTracker` to override `result`.
- Adjusted references in `Player` to interact with `result` through the interface.

SebasNadu committed:
refactor: change `Hand` constructor to accept initial cards and add size property

SebasNadu committed:
feat: implement game states for Blackjack including Hit, Stand, Bust, and Finished

SebasNadu committed:
refactor: clean up code formatting and add missing newlines in several files

Wordbe commented on README.md:
```diff
@@ -58,11 +63,18 @@
     - [x] Should be comma separated names
     - [x] Should contain a minimum of one player.
     - [x] The names should be unique.
+  - [ ] Should accept an amount for the Bet bigger than 0.
```
Check this please!

SebasNadu replied:
Sorry!🤦

Wordbe commented on src/main/kotlin/blackjack/model/Chips.kt:
```diff
@@ -0,0 +1,23 @@
+
+    operator fun plus(other: Chips): Chips = Chips(value + other.value)
+
+    override fun toString(): String = "${value.toInt()}"
```
If it's being used in the view, it would be better to place it within the view. 👀

Wordbe commented on src/main/kotlin/blackjack/model/Chips.kt:
```diff
@@ -0,0 +1,23 @@
+        private const val BLACKJACK_MULTIPLIER = 1.5
+        private const val ZERO = 0.0
+
+        fun zero(): Chips = Chips(ZERO)
```
How about creating Chips.ZERO?

Wordbe commented on src/main/kotlin/blackjack/model/Chips.kt:
```diff
@@ -0,0 +1,23 @@
+package blackjack.model
+
+@JvmInline
+value class Chips(val value: Double) {
```
Good idea, value class!

Wordbe commented on src/main/kotlin/blackjack/model/GameJudge.kt:
```diff
@@ -10,12 +10,14 @@ object GameJudge {
         player.recordResult(result)
         dealer.recordResult(result.inverse)
+        player.addProfit()
+        dealer.addProfit(-player.profit)
```
How about implementing a subtraction operation?

Wordbe commented on src/main/kotlin/blackjack/model/card/Card.kt:
```diff
@@ -1,6 +1,6 @@
 package blackjack.model.card
 
-data class Card(val suit: Suit, val rank: Rank) {
+data class Card(val suit: Suit, val rank: Rank) : Comparable<Card> {
```
Using Comparable, it's good idea. 👍

Wordbe commented on src/main/kotlin/blackjack/model/card/Hand.kt:
```diff
@@ -1,8 +1,11 @@
 package blackjack.model.card
 
-class Hand(private val dealtCards: MutableList<Card> = mutableListOf()) {
+class Hand(initCards: List<Card> = emptyList()) {
```
Do we need to keep both initCards and dealtCards? 🙂

SebasNadu replied:
I modified the constructor with initCards: List<Card> to improve safety and testability. This allows passing an immutable list when creating a Hand, which is then defensively copied into an internal MutableList (dealtCards).

This way, external code cannot mutate the original input list, and the internal state remains protected. It also makes tests easier to write, since I can initialize a hand with a predefined set of cards without risking accidental mutation of test data.

Wordbe replied:
So you're saying it's used as a `MutableList` only internally - that's a good idea.
Thanks for the explanation! 👍

Wordbe commented on src/main/kotlin/blackjack/model/participant/Dealer.kt:
```diff
@@ -21,14 +22,14 @@ class Dealer(name: String = "Dealer", internal val deck: Deck = Deck.generateADe
 
-    override fun toString(): String {
-        return name
+    fun addProfit(playerProfit: Chips) {
```
I’d prefer the object name and variable name to be the same — the domain terminology can get a bit confusing otherwise!

SebasNadu replied:
Ah, I see! You’re referring to the variable name playerProfit, right? That makes sense, I named it that way because the profit comes from the player. It’s the player’s profit that affects the dealer’s profit. But yes, I get how that can be confusing in the context.

What do you think about renaming it to:
`fun deductFromProfit(chips: Chips)`

I feel this better reflects the actual operation, and avoids any ambiguity about ownership.

Wordbe replied:
I'm glad it's clearer now. thank you! 😊

Wordbe commented on src/main/kotlin/blackjack/model/participant/Participant.kt:
```diff
@@ -1,29 +1,32 @@
 abstract class Participant(val name: String, protected val resultTracker: ResultTracker) {
     @Suppress("PropertyName")
-    protected val _hand = Hand()
+    protected val participantHand = Hand()
```
What was the reason for changing hand to participantHand?
Personally, I think the shorter name looks better!

SebasNadu replied:
We discussed this with Jason. The reason for renaming _hand to participantHand is to avoid confusion, when you see _hand in a subclass, it immediately looks like a private variable due to the underscore prefix. But it’s actually a protected variable inherited from the parent class. Since it’s not private, using _ can be misleading. Renaming it makes the visibility and origin clearer. We also agreed to reserve the underscore prefix strictly for private variables. Do you agree with this practice, or do you think that both protected and private class could be assigned with _?🤔🤔

Wordbe replied:
Yes, I agree that using `_` only for private properties makes sense. It's a common convention.
But what I was referring to wasn't `_hand`, but `hand` itself!

Ah, I see now. The inherited objects are using `hand`, which makes sense.
Thanks for the clarification! It’s totally fine to keep it as is. Appreciate the explanation!

Wordbe commented on src/main/kotlin/blackjack/model/participant/Player.kt:
```diff
@@ -1,11 +1,30 @@
+import blackjack.model.result.Result
 
 class Player(name: String) : Participant(name, PlayerResultTracker()) {
+    var bet: Chips = Chips.zero()
```
Is it possible to make bet and profit immutable—ideally without maintaining any internal state?

SebasNadu replied:
Yes, I think so, this is a nice idea. I'll check that! Thank you!

Wordbe commented on src/main/kotlin/blackjack/model/result/DealerResultTracker.kt:
```diff
@@ -1,6 +1,7 @@
 package blackjack.model.result
 
 class DealerResultTracker : ResultTracker {
+    override var result: Result? = null
```
What was the reason for allowing null?
Could we try refactoring it to avoid allowing null altogether?

SebasNadu replied:
The result property is only relevant for PlayerResultTracker. Initially, I kept it exclusively there. However, this led to a problem: in Participant, which can represent either a player or a dealer, I store a reference to ResultTracker.
When accessing the result in player-specific logic, I had to cast ResultTracker to PlayerResultTracker using as, just to access the result property.
To avoid this unsafe cast, I moved result to the common ResultTracker interface, and for the dealer, it’s simply initialized as null.

An alternative would be to avoid sharing the same ResultTracker type between player and dealer. We could use a PlayerResultTracker in Player and a DealerResultTracker in Dealer, and not reference them via a common Participant field.

I’m open to suggestions. Which do you think is more idiomatic: keeping a shared tracker interface with optional properties, or using distinct tracker fields per subclass?

Wordbe replied:
It seems like having a distinct tracker for each subclass is a good choice.
It improves type safety - no need for casting or null checks
and clearly expresses intent.
Thanks for thinking it through and implementing it! 🙏

Wordbe commented on src/main/kotlin/blackjack/state/Finished.kt:
```diff
@@ -0,0 +1,19 @@
+
+import blackjack.model.card.Card
+
+abstract class Finished : State {
```
Nice 🙂approach.

Wordbe commented on src/main/kotlin/blackjack/state/Hit.kt:
```diff
@@ -0,0 +1,20 @@
+
+class Hit(override val hand: Hand) : Running() {
+    init {
+        require(hand.size >= 2)
```
Please replace the magic numbers with constants, and update them everywhere else as well.

Wordbe commented on src/main/kotlin/blackjack/state/Running.kt:
```diff
@@ -0,0 +1,7 @@
+package blackjack.state
+
+abstract class Running : State {
+    override fun profit(money: Int): Double {
```
How about creating a Money object to wrap the primitive value? It could improve clarity and type safety. 🙂

Wordbe commented on src/main/kotlin/blackjack/state/State.kt:
```diff
@@ -0,0 +1,14 @@
+import blackjack.model.card.Card
+import blackjack.model.card.Hand
+
+interface State {
```
Was there a particular reason you chose an interface over an abstract class? 🙂

Wordbe reviewed (changes requested):
Hello Sebas!

Great job on completing the Blackjack mission and submitting your code! 🙌
It was impressive to see how you introduced State to manage the game status of participants.

I've left some comments with points that might be worth considering — please take a look and reflect any changes where possible.

Thank you! 😊

SebasNadu committed:
refactor: simplify Participant and Player result tracking

- Removed `resultTracker` from `Participant` and moved it directly into `Player`.
- Updated `Player` to explicitly handle result tracking using `PlayerResultTracker`.
- Replaced `dealer.addProfit` with `dealer.deductFromProfit` for clearer intent.
- Refactored `bet` and `profit` to use consistent `Chips.zero`.

SebasNadu committed:
feat: add subtraction support and update `Chips.zero` implementation

- Added `minus` operator to `Chips` for subtraction functionality.
- Updated `zero` from a function to a property for improved usage consistency.

SebasNadu committed:
refactor: update profit display to use integer values for clarity

- Modified `printResults` in `OutputView` to display profits as integers for both dealer and players.

SebasNadu committed:
docs: update README to mark bet amount validation as completed

SebasNadu committed:
refactor: remove redundant imports and add missing newlines

- Cleaned up unused imports across `Participant`, `Player`, and `Dealer`.
- Added missing newlines for consistent file formatting.

SebasNadu committed:
Merge remote-tracking branch 'origin/step2' into step2

# Conflicts:
#	src/main/kotlin/blackjack/model/result/ResultTracker.kt

SebasNadu commented:
Hi Wordbe! 👋
Thanks again for your feedback.
In this second iteration, I followed your suggestions: I fixed the issues you pointed out, renamed some methods and variables for clarity, and added the helpful methods you proposed.

Regarding the ResultTracker, I went with the second option to avoid a nullable result. Now, both Player and Dealer contain their own instances of PlayerResultTracker and DealerResultTracker, each with its own record() method. This slightly weakens the benefits of inheritance from Participant, but improves readability and eliminates the nullable concern.

Please ignore the state folder. I experimented with the State pattern but ultimately decided to stick with my initial approach for now. The current implementation doesn’t use any of the classes in that folder. Sorry for the confusion.

Thanks again for your time and insights, you’ve really helped me level up! 🚀🚀

Wordbe commented on src/main/kotlin/blackjack/model/result/PlayerResultTracker.kt:
```diff
@@ -1,13 +1,13 @@
 
     override fun toString(): String {
-        return lastResult?.description ?: "No result recorded."
+        return result?.description ?: "No result recorded."
```
The null operator is no longer needed.
Also, if `toString()` is being used for the View, consider moving that responsibility to the View layer instead. (Same feedback as before) 🙂

Wordbe commented on src/main/kotlin/blackjack/model/participant/Participants.kt:
```diff
@@ -1,15 +1,23 @@
         get() = dealer.getScore()
     val dealerName: String
         get() = dealer.name
     val dealerResult: String
         get() = dealer.result
```
Since it's no longer used, please remove it!

Wordbe reviewed (changes requested):
Hello Sebas!

Thank you for your great work on Blackjack mission Step 2.
Are you planning to merge the part you created with `State` into the main code? Please check on this.

Also, please review the comments I left and apply any changes you can.
Thanks! 😊

SebasNadu commented:
Yes, you're right. I'll create a new branch for the upcoming state feature and keep the step2 branch clean and ready for merging. Thanks a lot for your insights and the patience.

SebasNadu committed:
refactor: remove obsolete Blackjack state classes and simplify PlayerResultTracker

SebasNadu commented:
Hi Wordbe! 👋

I’ve removed the unused state folder from this branch, cleaned up the unused result property in Participants, and eliminated the unnecessary null check in PlayerResultTracker.

Thanks again for your thoughtful review, and for your patience with the small oversights. I’ll make sure to pay closer attention to those details going forward 🙏🙂

I hope the code is now ready to be merged! 🚀

SebasNadu committed:
refactor(participants): remove unused dealerResult property

Wordbe reviewed (approved):
Hi Sebas!
Thanks a lot for the cleanups. I really appreciate the attention to detail.
The code looks great now, and is in good shape for merging.

As you've already started doing, I really hope you take the time to apply the State pattern on your own. It's a great way to truly experience object-oriented design and the power of polymorphism.

Well done so far, and best of luck with your next mission! You've got this! 💪🚀
